src/python/pants/backend/go/lint/golangci_lint/rules.py

```diff
@@ -0,0 +1,187 @@
+
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        return tgt.get(SkipGolangciLintField).value
+
+
+class GolangciLintRequest(LintTargetsRequest):
+    field_set_type = GolangciLintFieldSet
+    name = GolangciLint.options_scope
+
+
+@rule
+async def partition_golangci_lint(
+    request: GolangciLintRequest.PartitionRequest[GolangciLintFieldSet], golangci_lint: GolangciLint
+) -> Partitions[GolangciLintFieldSet]:
+    return Partitions() if golangci_lint.skip else Partitions.single_partition(request.field_sets)
+
+
+@rule(desc="Lint with golangci-lint", level=LogLevel.DEBUG)
+async def run_golangci_lint(
+    request: GolangciLintRequest.SubPartition[GolangciLintFieldSet],
+    golangci_lint: GolangciLint,
+    goroot: GoRoot,
+    bash: BashBinary,
+    platform: Platform,
+    golang_subsystem: GolangSubsystem,
+) -> LintResult:
+    transitive_targets = await Get(
+        TransitiveTargets,
+        TransitiveTargetsRequest((field_set.address for field_set in request.elements)),
+    )
+
+    all_source_files_request = Get(
+        SourceFiles,
+        SourceFilesRequest(
+            tgt[SourcesField] for tgt in transitive_targets.closure if tgt.has_field(SourcesField)
+        ),
+    )
+
+    target_source_files_request = Get(
+        SourceFiles,
+        SourceFilesRequest(field_set.sources for field_set in request.elements),
+    )
+
+    downloaded_golangci_lint_request = Get(
+        DownloadedExternalTool,
+        ExternalToolRequest,
+        golangci_lint.get_request(platform),
+    )
+
+    config_files_request = Get(ConfigFiles, ConfigFilesRequest, golangci_lint.config_request())
+
+    (
+        target_source_files,
+        all_source_files,
+        downloaded_golangci_lint,
+        config_files,
+    ) = await MultiGet(
+        target_source_files_request,
+        all_source_files_request,
+        downloaded_golangci_lint_request,
+        config_files_request,
+    )
+
+    owning_go_mods = await MultiGet(
+        Get(OwningGoMod, OwningGoModRequest(field_set.address)) for field_set in request.elements
+    )
+
+    owning_go_mod_addresses = {x.address for x in owning_go_mods}
+
+    go_mod_infos = await MultiGet(
+        Get(GoModInfo, GoModInfoRequest(address)) for address in owning_go_mod_addresses
+    )
+
+    # golangci-lint requires a absolute path to a cache
+    golangci_lint_run_script = FileContent(
+        "__run_golangci_lint.sh",
+        textwrap.dedent(
+            f"""\
+            export GOROOT={goroot.path}
+            sandbox_root="$(/bin/pwd)"
+            export PATH="${{GOROOT}}/bin:${{PATH}}"
+            export GOPATH="${{sandbox_root}})/gopath"
+            export GOCACHE="${{sandbox_root}}/gocache"
+            export GOLANGCI_LINT_CACHE="$GOCACHE"
+            export CGO_ENABLED={1 if golang_subsystem.cgo_enabled else 0}
+            /bin/mkdir -p "$GOPATH" "$GOCACHE"
+            exec "$@"
+            """
+        ).encode("utf-8"),
```

Comment:
Would it be useful if `GoSdkProcess` supported not invoking `go` if a boolean option was set? (Instead of replicating some of the functionality of that runner script here.)

Also maybe consider setting the `CGO_ENABLED` value using the `env` field on `Process` similar to what is done with usages of `GoSdkProcess`?

Reply:
I can definitely set `CGO_ENABLED` in the env argument. However, the runner script needs to set `GOLANGCI_LINT_CACHE` based on the sandboxed `GOCACHE`, so `GoSdkProcess` would need a way to inject additional env variables as well.

I'll see what that looks like.

Reply:
I'm also fine with just landing the PR as-is and refactoring separately.

Reply:
I immediately hit a snag with extending `GoSdkProcess` myself. I don't understand enough about how `go_sdk_invoke_setup()` is called.

Reply:
Then no need to handle my comment. Barring any other comments, I'll merge this tomorrow.